### PR TITLE
New: Add id-length rule (fixes #2784)

### LINKED ISSUE #2784: A rule to enforce a minimum identifier name length
I propose adding a rule that checks any identifier names against a predefined minimum length (2 by default).

**The use case for the rule**: People tend to use variable names such as `e` or `x` which makes the code harder to read and less maintainable. Proposal is to add a rule that has the following properties:
 - Have a minimum length for property and variable names
 - Have a predefined, sane ignore list: `i`, `j`, `$` and `_` for now
 - Have the ability to add exceptions per expression type

This is a purely stylistic rule and I think is generic enough to be included.

I think the rule should be off by default since it may throw some people off who tend to use `x`, `y`, `a`, `b`, `e` as variable names extensively in their code.

A sample implementation which we started to use at Disqus can be found here: https://gist.github.com/BYK/837ca4e7c12d1ee1693c

I'll create a pull request if this looks good.

### PR DESCRIPTION
**Pull request by BYK**
Fixes #2784

### CODE REVIEW

File: tests/lib/rules/id-length.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    ],
+    invalid: [
+        { code: "var x = 1;", errors: [{ message: "Identifier name 'x' is too short. (< 2)", type: "Identifier"}] },
+        { code: "var x;", errors: [{ message: "Identifier name 'x' is too short. (< 2)", type: "Identifier"}] },
+        { code: "function x() {};", errors: [{ message: "Identifier name 'x' is too short. (< 2)", type: "Identifier"}] },
+        { code: "function xyz(a) {};", errors: [{ message: "Identifier name 'a' is too short. (< 2)", type: "Identifier"}] },
+        { code: "var obj = {a: 1, bc: 2};", errors: [{ message: "Identifier name 'a' is too short. (< 2)", type: "Identifier"}] },
+        { code: "try { blah(); } catch (e) { /* pass */ }", errors: [{ message: "Identifier name 'e' is too short. (< 2)", type: "Identifier"}] },
+        { code: "var handler = function (e) {};", errors: [{ message: "Identifier name 'e' is too short. (< 2)", type: "Identifier"}] },
+        { code: "var handler = function ($e) {};", errors: [{ message: "Identifier name '$e' is too short. (< 2)", type: "Identifier"}] },
```
Comment (BYK) [inline tests/lib/rules/id-length.js]: These type of errors might be a bit confusing since it says `$e` is too short (< 2), where it actually is two characters and we check the "bare" length after stripping prefixes and suffixes. I couldn't find a succinct enough error message to express this subtlety :(

File: conf/eslint.json [hunk 2]
```diff
@@ -137,6 +137,7 @@
         "guard-for-in": 0,
         "handle-callback-err": 0,
         "indent": 0,
+        "id-length": 0,
```
Comment (BYK) [inline conf/eslint.json]: `id-len` might be an even better name.

File: lib/rules/id-length.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                "type": "array",
+                "uniqueItems": true,
+                "items": {
+                    "enum": [
+                        "AssignmentExpression",
+                        "VariableDeclarator",
+                        "ObjectExpression",
+                        "FunctionExpression",
+                        "FunctionDeclaration",
+                        "CatchClause"
```
Comment (gyandeeps) [inline lib/rules/id-length.js]: Please mention this in docs also that these are the available options for `disableFor`. Makes it easier for people.

**Comment by gyandeeps** [comment]
Please format your commit message as per our guidelines: http://eslint.org/docs/developer-guide/contributing

**Comment by BYK** [comment]
@gyandeeps done! :)

File: docs/rules/id-length.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+This rule is aimed at increasing code readability and maintainability by enforcing an identifier length convention. It will warn on any type of identifier which doesn't conform to length limits (upper and lower). The rule ignores common prefixes and suffixes up to two chars of `_` and `$` in any combination.
+
+It has exceptions for commonly used single-char identifier names such as `i` or `$`. It also allows the programmers to silently by-pass this check by using "quoted" property names or calculated property access to allow potential server-side data requirements.
+
+The following patterns are considered warnings:
+
+```js
+// id-length: 1
+var x = 5;
```
Comment (nzakas) [inline docs/rules/id-length.md]: Can you add some blank lines in between each example here?

File: docs/rules/id-length.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+}
+var myObj = { a: 1 };
+```
+
+The following patterns are not considered warnings:
+
+```js
+// id-length: 1
+var num = 5;
+function _func() { return 42; }
```
Comment (nzakas) [inline docs/rules/id-length.md]: Same here

File: docs/rules/id-length.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+```
+
+### Options
+
+The `id-length` rule has no required options and has 4 optional ones that needs to be passed in a single options object:
+
+* **min** *(default: 2)*: The minimum number of characters an identifier name should be, after it is stripped from it is prefixes and suffixes
+* **max** *(default: Infinity)*: The maximum number of characters an identifier name should be, after it is stripped from it is prefixes and suffixes
+* **exceptions** *(default: ["$", "_", "i", "j"])*: An array of identifier names that the rule should not apply to
+* **disableFor** *(default: [])*: An array of node types that the rule should not be applied to. Available types are:
```
Comment (nzakas) [inline docs/rules/id-length.md]: Is this necessary? I'm generally not a big fan of requiring node types in configuration and I don't see a clear use case for this.

File: docs/rules/id-length.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+```js
+// id-length: 1  // minimum 2-chars
+var x = 5; // too short
+```
+
+## Rule Details
+
+This rule is aimed at increasing code readability and maintainability by enforcing an identifier length convention. It will warn on any type of identifier which doesn't conform to length limits (upper and lower). The rule ignores common prefixes and suffixes up to two chars of `_` and `$` in any combination.
+
+It has exceptions for commonly used single-char identifier names such as `i` or `$`. It also allows the programmers to silently by-pass this check by using "quoted" property names or calculated property access to allow potential server-side data requirements.
```
Comment (nzakas) [inline docs/rules/id-length.md]: We cannot prepopulate identifiers in this way as there would be no way to say that these should not be allowed. There need to be no exceptions by default.

**Comment by nzakas** [comment]
This is a good start. One of the challenging things about core rules is that they need to be consistently generic, and this rule in its current incarnation has a lot of gotchas.

First, I was under the impression that this would be just for variable and function names, but it looks like it's also for properties. I'm not sure it makes sense to bundle those together, and if it does, we need a way to separate them so people can have different rules for variables/functions as compared to properties.

Next, we can't have any special treatment for anything baked in. That means no predefined exceptions and no special treatment for prefixes. The length must always be treated as absolute.

Keep in mind, it's easy to add things we miss later, but it's nearly impossible to remove things once they're introduced, so we always start with very simple rules and add options as we find necessary vs. trying to figure out every use case ahead of time and ship a big thing.

So, I think we should strip this rule down to the minimum implementation as a first pass and wait for feedback before adding in all the bells and whistles.

**Comment by michaelficarra** [comment]
Agreed and well put, @nzakas.

File: docs/rules/id-length.md (hunk 7 above)
Comment (BYK) [inline docs/rules/id-length.md]: I almost agree but not including `i` as a default sounds like a crime. Are you 100% sure about this?

File: docs/rules/id-length.md (hunk 6 above)
Comment (BYK) [inline docs/rules/id-length.md]: I added this for the question you've raised:

>First, I was under the impression that this would be just for variable and function names, but it looks like it's also for properties. I'm not sure it makes sense to bundle those together, and if it does, we need a way to separate them so people can have different rules for variables/functions as compared to properties.

I personally don't use it but looks like we may wanna give people some "relaxation" options. May not be necessary for the first pass as you pointed out.

**Comment by BYK** [comment]
>First, I was under the impression that this would be just for variable and function names, but it looks like it's also for properties. I'm not sure it makes sense to bundle those together, and if it does, we need a way to separate them so people can have different rules for variables/functions as compared to properties.

I included properties because `exports` pattern allows people to assign any "variables" or "functions" to objects like this and they are essentially properties. That's why this rule became `id-length` instead of `var-length`. Not sure if we really need a way to define separate rules for these but it sounds double through a map: `"ExpressionType": { /* expression option overrides */ }`. Sounds a bit more complicated thoudh.

>Next, we can't have any special treatment for anything baked in. That means no predefined exceptions and no special treatment for prefixes. The length must always be treated as absolute.

I was thinking the same but when I started implementing this rule, I looked at other rules and took this idea from here: https://github.com/eslint/eslint/blob/master/lib/rules/camelcase.js#L53

>Keep in mind, it's easy to add things we miss later, but it's nearly impossible to remove things once they're introduced, so we always start with very simple rules and add options as we find necessary vs. trying to figure out every use case ahead of time and ship a big thing.

>So, I think we should strip this rule down to the minimum implementation as a first pass and wait for feedback before adding in all the bells and whistles.

Totally agree but I also think things should come with sensible defaults. Of course there are things we can strip but I want you to reconsider them under the light of my reasoning. Then I'll follow the path you suggest.

Sounds like a deal?

File: docs/rules/id-length.md (hunk 7 above)
Comment (nzakas) [inline docs/rules/id-length.md]: Yes

**Comment by nzakas** [comment]
I understand your points, however, my original comments stand as my opinion for moving forward. Sensible defaults are not universal (as we have learned the hard way over time), and once introduced, are hard to change. The best option is always to start with no default exceptions and wait for feedback to tell us what is or isn't painful enough to make a change.

**Comment by BYK** [comment]
>The best option is always to start with no default exceptions and wait for feedback to tell us what is or isn't painful enough to make a change.

Alright then. Changes are upcoming.

**Comment by BYK** [comment]
Updated:
* Make examples readable
* Remove default exceptions
* Remove prefix/suffix stripping
* Remove `disableFor`

File: docs/rules/id-length.md (hunk 7 above)
Comment (btmills) [inline docs/rules/id-length.md]: @nzakas Would default exceptions be acceptable if specifying an exceptions list overrode the defaults, allowing users to remove the default exceptions? A couple examples:

```js
{
    "id-length": 2, // Exceptions default to ["i", "j", "_", "$"]
    "id-length": [2, {
        "exceptions": [] // No exceptions
    }],
    "id-length": [2, {
        "exceptions": ["$"] // Only allow "$"
    }],
    "id-length": [2, {
        "exceptions": ["i", "j", "_", "$", "x"] // Defaults plus "x"
    }]
}
```

I'm proposing this because I like those defaults, though I completely agree it should be possible to override them. (Conveniently, the modification to support this would only require changing the empty fallback `[]` to include the defaults.)

File: docs/rules/id-length.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
+# Limit minimum and maximum length for identifiers (id-length)
+
+Identifier names such as `e`, `x`, `_t` or sometimes very long ones like `hashGeneratorResultOutputContainerObject` which usually makes the code harder to read and less maintainable. To prevent this, some code bases may enforce a minimum and/or maximum identifier length. (usually min 2-chars)
+
+```js
+// id-length: 1  // minimum 2-chars
```
Comment (nzakas) [inline docs/rules/id-length.md]: this should be "min", correct?

Comment (nzakas) [inline docs/rules/id-length.md]: Ooh I see, I think this is confusing as-is. I dying think people will understand that 1 is not a length. I'd specify the options throughout.

File: tests/lib/rules/id-length.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        "var obj = { abc: 1, de: 2 };",
+        "var obj = { 'a': 1, bc: 2 };",
+        "var obj = {}; obj['a'] = 2;",
+        "abc = d;",
+        "try { blah(); } catch (err) { /* pass */ }",
+        "var handler = function ($e) {};",
+        "var _a = 2",
+        "var _ad$$ = new $;",
+        "var xyz = new ΣΣ();",
+        { code: "var x = Foo(42)", args: [1, {"min": 1}] },
```
Comment (nzakas) [inline tests/lib/rules/id-length.js]: Can you use `options` instead of args? We are deprecating args. `options` is the same except you don't need the first number (that's not actually used anywhere)

**Comment by nzakas** [comment]
This looks good, just a couple notes.

What do people think about id-length vs. id-len?

**Comment by ilyavolodin** [comment]
I think `id-length` is short enough.

**Comment by BYK** [comment]
Updated and rebased. Sorry it took a while, life intervened :)

Not sure if you'd like the final version of docs though. If you can be more specific about how to improve it, it'd really help.

File: lib/rules/id-length.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
+/**
+ * @fileoverview Rule that warns when identifier names are shorter or longer
+ *               than the values provided in configuration.
+ * @author Burak Yigit Kaya aka BYK
```
Comment (ilyavolodin) [inline lib/rules/id-length.js]: Missing copyright

File: lib/rules/id-length.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = function (context) {
+    var options = context.options[0] || {};
+    var minLength = typeof options.min === "number" ? options.min : 2;
```
Comment (ilyavolodin) [inline lib/rules/id-length.js]: Don't need to check for number, schema is going to blow up if non-number is going to be passed in.

File: lib/rules/id-length.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    var maxLength = typeof options.max === "number" ? options.max : Infinity;
+    var exceptions = (Array.isArray(options.exceptions) ? options.exceptions : [])
```
Comment (ilyavolodin) [inline lib/rules/id-length.js]: Same with `Array.isArray`

File: docs/rules/id-length.md [hunk 13]
```diff
@@ -0,0 +1,84 @@
+# Limit minimum and maximum length for identifiers (id-length)
+
+Identifier names such as `e`, `x`, `_t` or sometimes very long ones like `hashGeneratorResultOutputContainerObject` which usually makes the code harder to read and less maintainable. To prevent this, some code bases may enforce a minimum and/or maximum identifier length. (usually min 2-chars)
```
Comment (ilyavolodin) [inline docs/rules/id-length.md]: I would restructure this sentence. It just doesn't flow.

File: lib/rules/id-length.js (hunk 10 above)
Comment (BYK) [inline lib/rules/id-length.js]: I don't want a copyright. There are rules (some by you) missing copyright :)

**Comment by BYK** [comment]
@ilyavolodin - updated!

File: lib/rules/id-length.js (hunk 10 above)
Comment (ilyavolodin) [inline lib/rules/id-length.js]: Those rules by me were created way before we started adding copyrights:-) But if you don't want to add one, that's perfectly fine.

**Comment by ilyavolodin** [comment]
There are some linting errors in the code. That's why builds are failing:
```
lib/rules/id-length.js
  15:37  error  Unexpected use of undefined  no-undefined
  16:37  error  Unexpected use of undefined  no-undefined
```

**Comment by BYK** [comment]
>There are some linting errors in the code. That's why builds are failing:

Oh boy, I updated this without running `npm test` lol. Sorry about that.

File: lib/rules/id-length.js (hunk 10 above)
Comment (BYK) [inline lib/rules/id-length.js]: I mean I want anyone to do whatever he or she wants with this code and a copyright implies that they need to consult to me or something right? That's why I'm resisting. If my understanding is not correct though, I'd add it especially for consistency :)

**Comment by BYK** [comment]
@ilyavolodin  - I did fix those `no-undefined` violations, rebased and now seeing many many lint failures. I'm guessing this is due to the recently landed changes around default rules.

Should I wait until they are fixed on master and rebase or is this okay to merge now?

File: lib/rules/id-length.js [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+ * @author Burak Yigit Kaya aka BYK
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = function (context) {
```
Comment (mysticatea) [inline lib/rules/id-length.js]: I guess new rule config for ESLint itself disallows spaces before the `(` of the argument list.

**Comment by ilyavolodin** [comment]
@BYK Master is error free currently. All of the 1.0.0 patches has already landed into master, so it looks like those errors are valid. It looks like all of them are because of the slight difference in coding style. We do not use spaces before function parents and it looks like you do.

**Comment by BYK** [comment]
@ilyavolodin - interesting, these didn't happen before. Also I was getting many errors on other files even including the `Makefile.js`. Let me try again.

**Comment by mysticatea** [comment]
Ah, maybe, it's related with #3005 
On Windows, `npm test` reports many errors by `spaced-comments`

**Comment by BYK** [comment]
@mysticatea oh yeah that's it. Thanks for the pointer!

**Comment by ilyavolodin** [comment]
I think this is good to merge after rebase.

**Comment by BYK** [comment]
Rebased again :)

**Comment by ilyavolodin** [comment]
@BYK Apparently code coverage for branches felt below the threshold.

File: lib/rules/id-length.js (hunk 10 above)
Comment (xlong328) [inline lib/rules/id-length.js]: xlong328@qq.com

File: tests/lib/rules/id-length.js [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+var eslint = require("../../../lib/eslint"),
+    ESLintTester = require("eslint-tester");
```
Comment (nzakas) [inline tests/lib/rules/id-length.js]: Can you switch this to be:

```js
ESLintTester = require('../../../lib/testers/eslint-tester')
```

File: lib/rules/id-length.js (hunk 10 above)
Comment (platinumazure) [inline lib/rules/id-length.js]: Not true that copyright automatically implies restriction of use of code. Copyright just means that this particular code snippet is definitely owned (intellectually) by you. It is possible to use a software license which asserts copyright but allows distribution and modification under reasonable conditions. Check out the project's license for more information.

Comment (nzakas) [inline lib/rules/id-length.js]: @byk Basically, you own own whatever code you write. The CLA says you are licensing that code to ESLint so we can include it under our license terms. It's not restricting anyone's ability to use the code, it's just asserting that you created it AND that's what gives you the right to allow us to use it.

Comment (BYK) [inline lib/rules/id-length.js]: Alright, adding it.

File: tests/lib/rules/id-length.js (hunk 15 above)
Comment (BYK) [inline tests/lib/rules/id-length.js]: Done.

**Comment by BYK** [comment]
Alright, rebased, made requested changes + improved the branch coverage.

**Comment by IanVS** [comment]
Thanks for your persistence, @BYK

**Comment by BYK** [comment]
>Thanks for your persistence, @BYK

Not gonna give up ;)

**Comment by nzakas** [comment]
Lgtm
